### PR TITLE
Support tokenization of special tokens for word_piece_tokenizer

### PR DESCRIPTION
**Pull request by abuelnasr0**
this PR enables the user to tokenize special tokens from text input as was suggested in keras-team/keras-nlp#1395.
The behaviour is the same as `ByteBairTokenizer` I think. also the models' tokenizers that use `WordPieceTokenizer` have the same behaviour as the models' tokenizers that use `BytePairTokenizer`.

### CODE REVIEW

**Comment by mattdangerw** [comment]
Thanks! This is a great feature.

I will think more on this next week. I like the overall approach of doing this on top of tf-text instead of inside the tf-text ops. But we may want to think about this more generally. Something like this as an end state...

- All tokenizer have a `special_tokens_in_strings=False` argument. If `True`, input strings can contain special tokens, which will be handled correctly.
- For pretrained tokenizers, this is setup automatically to respect the model's special tokens.
- For a base tokenizer, a special token list can be provided by the user.
- We do this eventually for all subword tokenizers--sentencepiece, byte pair, and word piece.

Main question here for me... Can we guarantee for all tokenizer types that this will work without messing with the "black box" tf-text op? How?

@abuelnasr0 let me know what you think! Once we have an overall plan, landing this incrementally (e.g. starting with word piece as you are here), sounds good!

File: keras_nlp/models/bert/bert_tokenizer.py
Comment (mattdangerw) [inline keras_nlp/models/bert/bert_tokenizer.py:102]: I think with this change, we might want to introduce a `special_tokens` property to the base class, and move this check, that all special tokens are the vocabulary, to the base class. WDYT?

**Review by mattdangerw — CHANGES_REQUESTED** [review]
OK, dropping some comments here for the whole set of PRs here.

Re `special_tokens` or `unsplittable_tokens`, I think what we are missing here is configurability. The goal is to be able to turn off or on this behavior not at the *base tokenizer* class but at the *model level* implementations. Basically, we want this...

```python
tokenizer = BertTokenizer.from_preset("bert_base_en")
tokenizer(["Bert uses [SEP] to mark sequence boundaries"])
# Tokenize as the tokens "[", "SE", "#P", "]".
tokenizer = BertTokenizer.from_preset("bert_base_en", special_tokens_in_strings=True)
tokenizer(["Bert uses [SEP] to mark sequence boundaries"])
# Tokenize as the tokens "[SEP]", token 102.
```

We want to give users a knob, that works in a cross model way, to either disallow special tokens in strings, or allow them (without even needing to know all the special tokens themselves).

If you are building a production system, you usually want this setting off. You don't want random string that happen to contain special tokens to mess with the number of segments you have.

If you are writing a guide, or building a prototype, you will often want to allow this. Gaining the ability to add special tokens in string space is very readable.

Does that make sense?

**Comment by abuelnasr0** [comment]
@mattdangerw Indeed! Thanks for that, it sounds rational, and it's a point of view that I was not aware of. I will address that for `WordPieceTokenizer` and `BytePairTokenizer`.

File: keras_nlp/models/bert/bert_tokenizer.py
Comment (abuelnasr0) [reply keras_nlp/models/bert/bert_tokenizer.py]: sounds good. I have added special tokens check inside `set_vocabulary` function. take a look at the latest changes.

**Comment by abuelnasr0** [comment]
@mattdangerw Check out last changes. I have add `special_tokens_in_strings` Arg for base tokenizer and model tokenizers. If all good, I will go ahead and do the same for `BytePairTokenizer` 
The default behaviour of `BytePairTokenizer` will change. where `BytePairTokenizer` handles special tokens correctly by default now, but I will change it to not handle special tokens correctly and only when `special_tokens_in_strings` is `True`, it should tokenize special tokens correctly.

**Comment by mattdangerw** [comment]
Thanks! Will take a look shortly!

> The default behaviour of BytePairTokenizer will change. where BytePairTokenizer handles special tokens correctly by default now, but I will change it to not handle special tokens correctly and only when special_tokens_in_strings is True

Yeah this ok I think. We would just want to call it out as a breaking change clearly in our next release. Thankfully the fix is quite clear, for anyone who wants the old behavior, pass the option.

File: keras_nlp/models/distil_bert/distil_bert_tokenizer.py
Comment (mattdangerw) [inline keras_nlp/models/distil_bert/distil_bert_tokenizer.py:120]: Maybe we can shorten these comments a bit?
`del config["special_tokens"]  # Not configurable; set in __init__.`

File: keras_nlp/tokenizers/word_piece_tokenizer.py
Comment (mattdangerw) [reply keras_nlp/tokenizers/word_piece_tokenizer.py:379]: maybe move this to a docstring comment on the method too?

File: keras_nlp/tokenizers/word_piece_tokenizer.py
Comment (mattdangerw) [inline keras_nlp/tokenizers/word_piece_tokenizer.py:167]: seems like you could shorten this logic a bit...
```python
if split_on_cjk:
    split_pattern = WHITESPACE_PUNCTUATION_AND_CJK_REGEX
    keep_split_pattern = PUNCTUATION_AND_CJK_REGEX
else:
    split_pattern = WHITESPACE_AND_PUNCTUATION_REGEX
    keep_split_pattern = PUNCTUATION_REGEX
if special_tokens_pattern is not None:
     split_pattern = r"|".join(special_tokens_pattern, split_pattern)
     keep_split_pattern = r"|".join(special_tokens_pattern, keep_split_pattern)
```
something like that

**Review by mattdangerw — COMMENTED** [review]
Nice! This looks good to me. Just a couple comments.

File: keras_nlp/tokenizers/word_piece_tokenizer.py
Comment (abuelnasr0) [reply keras_nlp/tokenizers/word_piece_tokenizer.py:167]: Thanks!

**Review by mattdangerw — APPROVED** [review]
lgtm! will pull in when CI is done

**Pull request merged**
